Connect to DuckDB

In [1]:
# Load the extension
%load_ext sql

There's a new jupysql version available (0.10.14), you're running 0.10.12. To upgrade: pip install jupysql --upgrade
Deploy Dash apps for free on Ploomber Cloud! Learn more: https://ploomber.io/s/signup


In [2]:
# Connect to DuckDB
%sql duckdb:///tpch.db

Connecting to 'duckdb:///tpch.db'

In [3]:
%config SqlMagic.displaylimit = 100

In [4]:
%%sql
-- Run a simple show tables
SELECT
  table_name
FROM
  information_schema.tables
WHERE
  table_schema = 'main'

Running query in 'duckdb:///tpch.db'

+------------+
| table_name |
+------------+
|  customer  |
|  lineitem  |
|   nation   |
|   orders   |
|    part    |
|  partsupp  |
|   region   |
|  supplier  |
+------------+

In [ ]:
# If you do not see any tables run the below command (after uncommeting it)
#! python setup.py

# [GET DATA] Use SELECT...FROM, LIMIT, WHERE, & ORDER BY to read the required data from tables

The most common use for querying is to read data in our tables. We can do this using a `SELECT ... FROM` statement, as shown below.



In [5]:
%%sql
SELECT
  *
FROM
  customer
LIMIT
  1;

Running query in 'duckdb:///tpch.db'

+-----------+--------------------+-------------------+-------------+-----------------+-----------+--------------+----------------------------------------------------------------+
| c_custkey |       c_name       |     c_address     | c_nationkey |     c_phone     | c_acctbal | c_mktsegment |                           c_comment                            |
+-----------+--------------------+-------------------+-------------+-----------------+-----------+--------------+----------------------------------------------------------------+
|     1     | Customer#000000001 | j5JsirBM9PsCy0O1m |      15     | 25-989-741-2988 |   711.56  |   BUILDING   | y final requests wake slyly quickly special accounts. blithely |
+-----------+--------------------+-------------------+-------------+-----------------+-----------+--------------+----------------------------------------------------------------+

In [6]:
%%sql
-- use * to specify all columns
SELECT
  *
FROM
  orders
LIMIT
  5;

Running query in 'duckdb:///tpch.db'

+------------+-----------+---------------+--------------+-------------+-----------------+-----------------+----------------+---------------------------------------------------------------------------+
| o_orderkey | o_custkey | o_orderstatus | o_totalprice | o_orderdate | o_orderpriority |     o_clerk     | o_shippriority |                                 o_comment                                 |
+------------+-----------+---------------+--------------+-------------+-----------------+-----------------+----------------+---------------------------------------------------------------------------+
|     1      |   36901   |       O       |  173665.47   |  1996-01-02 |      5-LOW      | Clerk#000000951 |       0        |                     ly express platelets. deposits acc                    |
|     2      |   78002   |       O       |   46929.18   |  1996-12-01 |     1-URGENT    | Clerk#000000880 |       0        |          ve the furiously fluffy dependencies. carefully regular          |
|     3      |   123314  |       F       |  193846.25   |  1993-10-14 |      5-LOW      | Clerk#000000955 |       0        |  after the asymptotes. instructions cajole after the foxes. carefully unu |
|     4      |   136777  |       O       |   32151.78   |  1995-10-11 |      5-LOW      | Clerk#000000124 |       0        |   st the furiously bold pinto beans. furiously pending theodolites cajol  |
|     5      |   44485   |       F       |  144659.20   |  1994-07-30 |      5-LOW      | Clerk#000000925 |       0        |           onic requests. carefully daring foxes among the carefu          |
+------------+-----------+---------------+--------------+-------------+-----------------+-----------------+----------------+---------------------------------------------------------------------------+

However, running a `SELECT ... FROM` statement can cause issues when the data set is extensive.


In [7]:
%%sql
-- use * to specify all columns
SELECT
  *
FROM
  orders
LIMIT
  5;

Running query in 'duckdb:///tpch.db'

+------------+-----------+---------------+--------------+-------------+-----------------+-----------------+----------------+---------------------------------------------------------------------------+
| o_orderkey | o_custkey | o_orderstatus | o_totalprice | o_orderdate | o_orderpriority |     o_clerk     | o_shippriority |                                 o_comment                                 |
+------------+-----------+---------------+--------------+-------------+-----------------+-----------------+----------------+---------------------------------------------------------------------------+
|     1      |   36901   |       O       |  173665.47   |  1996-01-02 |      5-LOW      | Clerk#000000951 |       0        |                     ly express platelets. deposits acc                    |
|     2      |   78002   |       O       |   46929.18   |  1996-12-01 |     1-URGENT    | Clerk#000000880 |       0        |          ve the furiously fluffy dependencies. carefully regular          |
|     3      |   123314  |       F       |  193846.25   |  1993-10-14 |      5-LOW      | Clerk#000000955 |       0        |  after the asymptotes. instructions cajole after the foxes. carefully unu |
|     4      |   136777  |       O       |   32151.78   |  1995-10-11 |      5-LOW      | Clerk#000000124 |       0        |   st the furiously bold pinto beans. furiously pending theodolites cajol  |
|     5      |   44485   |       F       |  144659.20   |  1994-07-30 |      5-LOW      | Clerk#000000925 |       0        |           onic requests. carefully daring foxes among the carefu          |
+------------+-----------+---------------+--------------+-------------+-----------------+-----------------+----------------+---------------------------------------------------------------------------+

In [8]:
%%sql
-- use column c_names only to read data from those columns
SELECT
  o_orderkey,
  o_totalprice
FROM
  orders
LIMIT
  5;

Running query in 'duckdb:///tpch.db'

+------------+--------------+
| o_orderkey | o_totalprice |
+------------+--------------+
|     1      |  173665.47   |
|     2      |   46929.18   |
|     3      |  193846.25   |
|     4      |   32151.78   |
|     5      |  144659.20   |
+------------+--------------+


We can use the `WHERE` clause if we want to get the rows that match specific criteria. We can specify one or more filters within the `WHERE` clause. 

The `WHERE` clause with more than one filter can use combinations of `AND` and `OR` criteria to combine the filter criteria, as shown below.



In [9]:
%%sql
-- all customer rows that have c_nationkey = 20
SELECT
  *
FROM
  customer
WHERE
  c_nationkey = 20
LIMIT
  10;

Running query in 'duckdb:///tpch.db'

+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+--------------------------------------------------------------------------------------------------------------------+
| c_custkey |       c_name       |                c_address                 | c_nationkey |     c_phone     | c_acctbal | c_mktsegment |                                                     c_comment                                                      |
+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+--------------------------------------------------------------------------------------------------------------------+
|     6     | Customer#000000006 |    g1s,pzDenUEBW3O,2 pxu0f9n2g64rJrt5E   |      20     | 30-114-968-4951 |  7638.57  |  AUTOMOBILE  |       quickly silent asymptotes are slyly regular excuses. instructions wake furiously? quickly bold courts p      |
|     81    | Customer#000000081 |            9jUFbrThIIeoUNd8 9            |      20     | 30-165-277-3269 |  2023.71  |   BUILDING   |                 s against the ironic packages haggle carefully above the slyly express pinto beans                 |
|    100    | Customer#000000100 |              MBy6qq3OEGpV4u              |      20     | 30-749-445-4907 |  9889.89  |  FURNITURE   |                   dazzle carefully furiously final foxes. express, ironic packages among the qui                   |
|    210    | Customer#000000210 |          ,XOlfSzkZDAkm96adR41j,          |      20     | 30-876-248-9750 |  7250.14  |  HOUSEHOLD   |                                       es cajole bravely across the blithely                                        |
|    223    | Customer#000000223 |          MyQxUcG0P QCetmG00GlF           |      20     | 30-193-643-1517 |  7476.20  |   BUILDING   |      xcuses. silent theodolites across the carefully bold excuses sleep ironic, final courts. regular excuses      |
|    228    | Customer#000000228 | rZ1wxvHNByT71bUJWZjXMDROzlAch6FVu,dj8Zfq |      20     | 30-435-915-1603 |  6868.12  |  FURNITURE   |                                           es. blithely permanent sentim                                            |
|    247    | Customer#000000247 |            eSAW4XaakYFj2WToKU            |      20     | 30-151-905-3513 |  8495.92  |  HOUSEHOLD   |                tes nag according to the blithe, even packages. sometimes unusual packages integrate                |
|    278    | Customer#000000278 |             XHAfHlrYQM3elmhJ             |      20     | 30-445-570-5841 |  7621.56  |   BUILDING   | ely unusual accounts. stealthily special instructions affix blithely. regular, ironic packages sleep even platelet |
|    285    | Customer#000000285 |   rB6fTQKle64k3MvCCatad8DfMgR5OZA G4r    |      20     | 30-235-130-1313 |  7276.72  |  FURNITURE   |         slyly according to the blithely special instructions. ironic ideas against the blithely furious pac        |
|    321    | Customer#000000321 |     LX0SKs3jqo9wH1yixIdGWp2ItclDiuL      |      20     | 30-114-675-9153 |  7718.77  |  FURNITURE   |                ng the final, bold requests. furiously regular accounts inside the furiously pending                |
+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+--------------------------------------------------------------------------------------------------------------------+

In [10]:
%%sql
-- all customer rows that have c_nationkey = 20 and c_acctbal > 1000
SELECT
  *
FROM
  customer
WHERE
  c_nationkey = 20
  AND c_acctbal > 1000
LIMIT
  10;

Running query in 'duckdb:///tpch.db'

+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+--------------------------------------------------------------------------------------------------------------------+
| c_custkey |       c_name       |                c_address                 | c_nationkey |     c_phone     | c_acctbal | c_mktsegment |                                                     c_comment                                                      |
+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+--------------------------------------------------------------------------------------------------------------------+
|     6     | Customer#000000006 |    g1s,pzDenUEBW3O,2 pxu0f9n2g64rJrt5E   |      20     | 30-114-968-4951 |  7638.57  |  AUTOMOBILE  |       quickly silent asymptotes are slyly regular excuses. instructions wake furiously? quickly bold courts p      |
|     81    | Customer#000000081 |            9jUFbrThIIeoUNd8 9            |      20     | 30-165-277-3269 |  2023.71  |   BUILDING   |                 s against the ironic packages haggle carefully above the slyly express pinto beans                 |
|    100    | Customer#000000100 |              MBy6qq3OEGpV4u              |      20     | 30-749-445-4907 |  9889.89  |  FURNITURE   |                   dazzle carefully furiously final foxes. express, ironic packages among the qui                   |
|    210    | Customer#000000210 |          ,XOlfSzkZDAkm96adR41j,          |      20     | 30-876-248-9750 |  7250.14  |  HOUSEHOLD   |                                       es cajole bravely across the blithely                                        |
|    223    | Customer#000000223 |          MyQxUcG0P QCetmG00GlF           |      20     | 30-193-643-1517 |  7476.20  |   BUILDING   |      xcuses. silent theodolites across the carefully bold excuses sleep ironic, final courts. regular excuses      |
|    228    | Customer#000000228 | rZ1wxvHNByT71bUJWZjXMDROzlAch6FVu,dj8Zfq |      20     | 30-435-915-1603 |  6868.12  |  FURNITURE   |                                           es. blithely permanent sentim                                            |
|    247    | Customer#000000247 |            eSAW4XaakYFj2WToKU            |      20     | 30-151-905-3513 |  8495.92  |  HOUSEHOLD   |                tes nag according to the blithe, even packages. sometimes unusual packages integrate                |
|    278    | Customer#000000278 |             XHAfHlrYQM3elmhJ             |      20     | 30-445-570-5841 |  7621.56  |   BUILDING   | ely unusual accounts. stealthily special instructions affix blithely. regular, ironic packages sleep even platelet |
|    285    | Customer#000000285 |   rB6fTQKle64k3MvCCatad8DfMgR5OZA G4r    |      20     | 30-235-130-1313 |  7276.72  |  FURNITURE   |         slyly according to the blithely special instructions. ironic ideas against the blithely furious pac        |
|    321    | Customer#000000321 |     LX0SKs3jqo9wH1yixIdGWp2ItclDiuL      |      20     | 30-114-675-9153 |  7718.77  |  FURNITURE   |                ng the final, bold requests. furiously regular accounts inside the furiously pending                |
+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+--------------------------------------------------------------------------------------------------------------------+

In [11]:
%%sql
-- all customer rows that have c_nationkey = 20 or c_acctbal > 1000
SELECT
  *
FROM
  customer
WHERE
  c_nationkey = 20
  OR c_acctbal > 1000
LIMIT
  10;

Running query in 'duckdb:///tpch.db'

+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+--------------------------------------------------------------------------------------------------------------------+
| c_custkey |       c_name       |                c_address                 | c_nationkey |     c_phone     | c_acctbal | c_mktsegment |                                                     c_comment                                                      |
+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+--------------------------------------------------------------------------------------------------------------------+
|     6     | Customer#000000006 |    g1s,pzDenUEBW3O,2 pxu0f9n2g64rJrt5E   |      20     | 30-114-968-4951 |  7638.57  |  AUTOMOBILE  |       quickly silent asymptotes are slyly regular excuses. instructions wake furiously? quickly bold courts p      |
|     81    | Customer#000000081 |            9jUFbrThIIeoUNd8 9            |      20     | 30-165-277-3269 |  2023.71  |   BUILDING   |                 s against the ironic packages haggle carefully above the slyly express pinto beans                 |
|    100    | Customer#000000100 |              MBy6qq3OEGpV4u              |      20     | 30-749-445-4907 |  9889.89  |  FURNITURE   |                   dazzle carefully furiously final foxes. express, ironic packages among the qui                   |
|    210    | Customer#000000210 |          ,XOlfSzkZDAkm96adR41j,          |      20     | 30-876-248-9750 |  7250.14  |  HOUSEHOLD   |                                       es cajole bravely across the blithely                                        |
|    223    | Customer#000000223 |          MyQxUcG0P QCetmG00GlF           |      20     | 30-193-643-1517 |  7476.20  |   BUILDING   |      xcuses. silent theodolites across the carefully bold excuses sleep ironic, final courts. regular excuses      |
|    228    | Customer#000000228 | rZ1wxvHNByT71bUJWZjXMDROzlAch6FVu,dj8Zfq |      20     | 30-435-915-1603 |  6868.12  |  FURNITURE   |                                           es. blithely permanent sentim                                            |
|    247    | Customer#000000247 |            eSAW4XaakYFj2WToKU            |      20     | 30-151-905-3513 |  8495.92  |  HOUSEHOLD   |                tes nag according to the blithe, even packages. sometimes unusual packages integrate                |
|    278    | Customer#000000278 |             XHAfHlrYQM3elmhJ             |      20     | 30-445-570-5841 |  7621.56  |   BUILDING   | ely unusual accounts. stealthily special instructions affix blithely. regular, ironic packages sleep even platelet |
|    285    | Customer#000000285 |   rB6fTQKle64k3MvCCatad8DfMgR5OZA G4r    |      20     | 30-235-130-1313 |  7276.72  |  FURNITURE   |         slyly according to the blithely special instructions. ironic ideas against the blithely furious pac        |
|    321    | Customer#000000321 |     LX0SKs3jqo9wH1yixIdGWp2ItclDiuL      |      20     | 30-114-675-9153 |  7718.77  |  FURNITURE   |                ng the final, bold requests. furiously regular accounts inside the furiously pending                |
+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+--------------------------------------------------------------------------------------------------------------------+

In [12]:
%%sql
-- all customer rows that have (c_nationkey = 20 and c_acctbal > 1000) or rows that have c_nationkey = 11
SELECT
  *
FROM
  customer
WHERE
  (
    c_nationkey = 20
    AND c_acctbal > 1000
  )
  OR c_nationkey = 11
LIMIT
  10;

Running query in 'duckdb:///tpch.db'

+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+---------------------------------------------------------------------------------------------------------+
| c_custkey |       c_name       |                c_address                 | c_nationkey |     c_phone     | c_acctbal | c_mktsegment |                                                c_comment                                                |
+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+---------------------------------------------------------------------------------------------------------+
|     52    | Customer#000000052 |       UracAlAA8tSHL5V,poTZIOjh8o,        |      11     | 21-186-284-5998 |  5630.28  |  HOUSEHOLD   |                ts boost. carefully express waters across the blithely regular foxes inte                |
|     84    | Customer#000000084 |   GB3sUmv RRXV DPzeOSbGxMIF9Z4Eq9 rop    |      11     | 21-546-818-3802 |  5174.71  |  FURNITURE   |          ounts. blithely express theodolites nag carefully ironic pinto beans. carefully final          |
|    131    | Customer#000000131 |      ItdUFrHPZlzjZ, fo03sG4topAKTV       |      11     | 21-840-210-3572 |  8595.53  |  HOUSEHOLD   |              ly final Tiresias. slyly permanent theodolites cajole quickly. carefully unus              |
|    134    | Customer#000000134 |        6I1TTaoG7bbiogCqRcptG6BYme        |      11     | 21-200-159-5932 |  4608.90  |   BUILDING   |                                  ly regular dolphins haggle blithely.                                   |
|    148    | Customer#000000148 |          qJ8bFn4kwiit7RzwGrwo5m          |      11     | 21-562-498-6636 |  2135.60  |  HOUSEHOLD   | e carefully pending ideas detect slyly along the furiously special excuses. instructions use carefully  |
|    190    | Customer#000000190 | mY30kK8AfsTGrx,L4zI   QlQnnmCUxikyc8QcZ7 |      11     | 21-730-373-8193 |  1657.46  |  AUTOMOBILE  |                                y even packages engage furiously pending p                               |
|    219    | Customer#000000219 | N7HIf0,MtChIqMY5aCsXnBbKYGw8fMdxY3Ywtug  |      11     | 21-159-138-6090 |  9858.57  |  AUTOMOBILE  |                                      carefully pending instructions                                     |
|    264    | Customer#000000264 |              ZWqGIvPWJAC9n               |      11     | 21-881-683-3829 |  3195.83  |  MACHINERY   |                                     egular dependencies wake reques                                     |
|    280    | Customer#000000280 |              YLnIkFdUWmE,MI              |      11     | 21-537-461-3965 |  3952.84  |   BUILDING   |      yly quickly ruthless accounts. unusual deposits solve furiously across the requests. carefull      |
|    292    | Customer#000000292 |              Ko3JXxuoWxIMR               |      11     | 21-457-910-2923 |  2975.43  |  HOUSEHOLD   |   ecial foxes play carefully silent deposits. quickly regular packages need to nag according to the th  |
+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+---------------------------------------------------------------------------------------------------------+


We can combine multiple filter clauses, as seen above. We have seen examples of equals (`=`) and greater than (`>`) conditional operators. There are 6 **conditional operators**, they are

1. **<** Less than
2. **>** Greater than
3. **<=** Less than or equal to
4. **>=** Greater than or equal to
5. **=** Equal
6. **<>** and **!=** both represent Not equal (some DBs only support one of these)

Additionally, for string types, we can make **pattern matching with `like` condition**. In a `like` condition, a `_` means any single character, and `%` means zero or more characters, for example.




In [13]:
%%sql
-- all customer rows where the c_name has a 381 in it
SELECT
  *
FROM
  customer
WHERE
  c_name LIKE '%381%';

Running query in 'duckdb:///tpch.db'

+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+---------------------------------------------------------------------------------------------------------------------+
| c_custkey |       c_name       |                c_address                 | c_nationkey |     c_phone     | c_acctbal | c_mktsegment |                                                      c_comment                                                      |
+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+---------------------------------------------------------------------------------------------------------------------+
|    381    | Customer#000000381 |             wXs5zN2nPHqPsfFO             |      5      | 15-860-208-7093 |  9931.71  |   BUILDING   |                           ithely along the regular, regular theodolites. fluffily pending                           |
|    1381   | Customer#000001381 |  kAgLl7nUiPStCleWOiKevH3QAOhqtg9dVvrdN   |      22     | 32-418-900-6494 |   367.82  |   BUILDING   |                                      posits sleep carefully around the slyly e                                      |
|    2381   | Customer#000002381 | z7B43DZ7RGlkgEi3YaXfy,Aw2SZepYurvII41Do  |      5      | 15-493-990-8133 |   412.99  |  FURNITURE   |                              ul requests use slyly quickly even deposits. slyly pending                             |
|    3381   | Customer#000003381 |          03jULkpVTm92eKW24meIj           |      13     | 23-441-750-5088 |  2473.54  |  AUTOMOBILE  |                          er the carefully bold multipliers doze blithely along the furiousl                         |
|    3810   | Customer#000003810 |            hlRTIO4e4HNahc8A D            |      18     | 28-881-994-8196 |  9906.80  |  FURNITURE   |                                           bold requests after the furiousl                                          |
|    3811   | Customer#000003811 |           b6vEJqifAgSbGhzTwTz            |      22     | 32-962-997-2221 |  5697.04  |  FURNITURE   |                          he carefully special packages. regular deposits sleep blithely bl                          |
|    3812   | Customer#000003812 |                HGYp5dZtlA                |      14     | 24-653-654-5032 |  4204.53  |  FURNITURE   |                                          y ironic requests believe blithely                                         |
|    3813   | Customer#000003813 |      Aeky0En0JO5V1zRgFZ9EvCcBWaTmW       |      6      | 16-983-191-7833 |  -494.03  |  HOUSEHOLD   |                                   rding to the express foxes. bold platelets main                                   |
|    3814   | Customer#000003814 |       FQ3lWCA3znooc3S SmDCfwqdn4R9       |      20     | 30-833-732-5401 |  -207.83  |  AUTOMOBILE  |                                        ounts alongside of the fluffily pendin                                       |
|    3815   | Customer#000003815 |          S5SIUeDCuVOKRTZqZ5M4CC          |      19     | 29-968-870-7672 |  2887.99  |  FURNITURE   |                                     ccounts. fluffily bold requests sleep furio                                     |
|    3816   | Customer#000003816 |                kFLu9DYSXH                |      6      | 16-806-750-3968 |  4353.05  |  AUTOMOBILE  |                 g the carefully ironic asymptotes. furiously ironic foxes haggle. fluffily bold pack                |
|    3817   | Customer#000003817 |  IWp4rgxgRxvUeHK2xtX8SEfGYmHYrCCpbBwXm   |      16     | 26-142-669-5787 |  9766.42  |  AUTOMOBILE  |                       es until the bold instructions wake quickly carefully unusual packages;                       |
|    3818   | Customer#000003818 | yrvC8HiUJVCsu7MyFHcyTcmAP0NgEWnoXQgTAz36 |      7      | 17-983-218-3716 |   346.99  |  AUTOMOBILE  |               theodolites. special ins

In [14]:
%%sql
-- all customer rows where the c_name ends with a 381
SELECT
  *
FROM
  customer
WHERE
  c_name LIKE '%381';

Running query in 'duckdb:///tpch.db'

+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+---------------------------------------------------------------------------------------------------------------------+
| c_custkey |       c_name       |                c_address                 | c_nationkey |     c_phone     | c_acctbal | c_mktsegment |                                                      c_comment                                                      |
+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+---------------------------------------------------------------------------------------------------------------------+
|    381    | Customer#000000381 |             wXs5zN2nPHqPsfFO             |      5      | 15-860-208-7093 |  9931.71  |   BUILDING   |                           ithely along the regular, regular theodolites. fluffily pending                           |
|    1381   | Customer#000001381 |  kAgLl7nUiPStCleWOiKevH3QAOhqtg9dVvrdN   |      22     | 32-418-900-6494 |   367.82  |   BUILDING   |                                      posits sleep carefully around the slyly e                                      |
|    2381   | Customer#000002381 | z7B43DZ7RGlkgEi3YaXfy,Aw2SZepYurvII41Do  |      5      | 15-493-990-8133 |   412.99  |  FURNITURE   |                              ul requests use slyly quickly even deposits. slyly pending                             |
|    3381   | Customer#000003381 |          03jULkpVTm92eKW24meIj           |      13     | 23-441-750-5088 |  2473.54  |  AUTOMOBILE  |                          er the carefully bold multipliers doze blithely along the furiousl                         |
|    4381   | Customer#000004381 |       MIQXH5W6Zsup5cVYfCtWupiJtgi        |      2      | 12-570-797-1472 |  2542.55  |  HOUSEHOLD   |                                      r deposits. carefully even packages along                                      |
|    5381   | Customer#000005381 |      bXQ,KuigJB1nASXN73PDwNOvXCIkp5      |      5      | 15-700-184-7619 |  4130.88  |  MACHINERY   |                                es. carefully ironic ideas sleep blithely about the i                                |
|    6381   | Customer#000006381 |          BKfk07DtN45gg2w4mMUK1           |      7      | 17-877-502-9214 |  7346.88  |  HOUSEHOLD   |        inal asymptotes boost. bold, ironic requests are along the regular, special packages. pending account        |
|    7381   | Customer#000007381 |    yq7RXRmclCUi6wJspelKaEWSJ  TfycLah    |      20     | 30-666-139-1602 |   73.39   |   BUILDING   | fluffily special requests are about the fluffily unusual foxes. final frets are slyly fluffily final deposits. even |
|    8381   | Customer#000008381 |  7kbg8wegbgGmgiW8OQ4SbJ8colXl6rpBmHudJ   |      0      | 10-177-308-9094 |  6674.59  |  AUTOMOBILE  |      uests against the carefully bold excuses sleep blithely slyly final instructions; unusual requests about       |
|    9381   | Customer#000009381 |              BhXODcEOpwNg6               |      17     | 27-708-588-6706 |  4788.15  |  HOUSEHOLD   |            sual hockey players use above the final packages. quickly ironic excuses sleep. slyly final pa           |
|   10381   | Customer#000010381 |               RLDSr5JOMK45               |      15     | 25-294-314-6077 |   106.15  |  FURNITURE   |                            r the furiously unusual pinto beans. blithely ironic pinto be                            |
|   11381   | Customer#000011381 |             cHcQKKDGnWLbq48O             |      7      | 17-652-294-7849 |  1959.81  |  MACHINERY   |                                   structions. slyly ironic requests play furiously                                  |
|   12381   | Customer#000012381 |     hPUYchVeqkeLC1QCsGJXF5Fj6QDQhkd      |      13     | 23-989-969-2819 |  1172.56  |  HOUSEHOLD   |                              cies cajo

In [15]:
%%sql
-- all customer rows where the c_name starts with a 381
SELECT
  *
FROM
  customer
WHERE
  c_name LIKE '381%';

Running query in 'duckdb:///tpch.db'

+-----------+--------+-----------+-------------+---------+-----------+--------------+-----------+
| c_custkey | c_name | c_address | c_nationkey | c_phone | c_acctbal | c_mktsegment | c_comment |
+-----------+--------+-----------+-------------+---------+-----------+--------------+-----------+
+-----------+--------+-----------+-------------+---------+-----------+--------------+-----------+

In [16]:
%%sql
-- all customer rows where the c_name has a combination of any character and 9 and 1
SELECT
  *
FROM
  customer
WHERE
  c_name LIKE '%_91%';

Running query in 'duckdb:///tpch.db'

+-----------+--------------------+-----------------------------------------+-------------+-----------------+-----------+--------------+----------------------------------------------------------------------------------------------------------------------+
| c_custkey |       c_name       |                c_address                | c_nationkey |     c_phone     | c_acctbal | c_mktsegment |                                                      c_comment                                                       |
+-----------+--------------------+-----------------------------------------+-------------+-----------------+-----------+--------------+----------------------------------------------------------------------------------------------------------------------+
|     91    | Customer#000000091 |  9Sce2m BjvDdjQkqMx8UnrUsJkk1IBAvZPTsA  |      8      | 18-239-400-3677 |  4643.14  |  AUTOMOBILE  |               yly ironic foxes lose slyly pending asymptotes. slyly final theodolites nag blithely ar                |
|    191    | Customer#000000191 |            cZMo3 b4GwZtUmdbw            |      16     | 26-811-707-6869 |  2945.16  |   BUILDING   |                                         daringly quickly ironic foxes. care                                          |
|    291    | Customer#000000291 |             2FfdPluDa2fxPaRh            |      8      | 18-657-656-2318 |  4261.68  |  HOUSEHOLD   |  ld deposits. regularly ironic pinto beans cajole permanently furiously express packages. regular, unusual sheaves   |
|    391    | Customer#000000391 |           BZ,850WgpZ0YSFs79Sb           |      11     | 21-604-451-4462 |  4801.30  |  HOUSEHOLD   |               tions wake about the blithely final instructions. excuses sleep regular requests. slyly                |
|    491    | Customer#000000491 |        AXsbcyMDujG,CAiEu4FmufbZ1k       |      0      | 10-856-259-7548 |   785.37  |  AUTOMOBILE  |                                ly final, even hockey players. carefully final ideas w                                |
|    591    | Customer#000000591 |            wkmTqEmyI3UOEoG3q            |      20     | 30-584-309-7885 |  6344.66  |  MACHINERY   |       xpress deposits. slyly ironic ideas haggle: daringly even requests after the quickly final ideas boost q       |
|    691    | Customer#000000691 |          0aGn3Vcf6ZKi82ogENfnso         |      16     | 26-741-688-4189 |  9566.15  |  MACHINERY   |             ven packages cajole fluffily fluffily unusual frays. ironic excuses sleep furiously. regular             |
|    791    | Customer#000000791 |           Y14aVvMuDDgnmEuCEPK           |      13     | 23-575-775-4059 |  3694.81  |  HOUSEHOLD   |                                beans use carefully furiously regular deposits. slyly                                 |
|    891    | Customer#000000891 |            r4,EU38BM0qdbjwqH            |      11     | 21-439-958-7518 |  6032.18  |  FURNITURE   |                                        ong the quickly quick patterns. slyly                                         |
|    910    | Customer#000000910 |              bKS7h8o7ZEiRj              |      9      | 19-899-463-4292 |  5794.69  |   BUILDING   |                 silent deposits are. blithely final foxes cajole slyly according to the furiously re                 |
|    911    | Customer#000000911 |        68,LIR0Fhraw6LUTrFW6Ew05u        |      13     | 23-121-746-7339 |  6364.60  |   BUILDING   |     uches. blithely silent requests haggle quick, bold accounts. carefully bold pinto beans according to the no      |
|    912    | Customer#000000912 |             OaqZYembag,qVEwo            |      14     | 24-348-437-3105 |  3861.36  |  FURNITURE   |                                       s sleep slyly along the regular, even i                                        |
|    913    | Customer#000000913 |             RCJd97HzePqdmZqv            |      14     | 24-500-946-3315 |  6935.16  |  FURNITURE   |                     symptotes. platelet

We can also filter for more than one value using `IN` and `NOT IN`.


In [17]:
%%sql
-- all customer rows which have c_nationkey = 10 or c_nationkey = 20
SELECT
  *
FROM
  customer
WHERE
  c_nationkey IN (10, 20);

Running query in 'duckdb:///tpch.db'

+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+----------------------------------------------------------------------------------------------------------------------+
| c_custkey |       c_name       |                c_address                 | c_nationkey |     c_phone     | c_acctbal | c_mktsegment |                                                      c_comment                                                       |
+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+----------------------------------------------------------------------------------------------------------------------+
|     16    | Customer#000000016 |           P2IQMff18ercaYrO,40            |      10     | 20-781-609-3107 |  4681.03  |  FURNITURE   |                         ests cajole. pinto beans detect slyly. final packages cajole slyly                           |
|     41    | Customer#000000041 |     jeREsFtCuMqEwdvTFqTkY2NzGRYDG1m      |      10     | 20-917-711-4011 |   270.95  |  HOUSEHOLD   |       uctions wake carefully pending deposits: pinto beans along the carefully final deposits sleep blithely a       |
|     49    | Customer#000000049 |  PdKqM4TlA OLTjaeRmvH7QWDu80USfslgqutF   |      10     | 20-908-631-4424 |  4573.94  |  FURNITURE   |                           quests haggle! furiously unusual theodolites cajole carefully. t                           |
|     55    | Customer#000000055 |  ti9p9XgdmFsjsQI6XQrISDUMFAusnmKS SBoCE  |      10     | 20-180-440-8525 |  4572.11  |  MACHINERY   |    dolites. bold instructions wake fluffily regular ideas. regular theodolites are furiously carefully unusual ac    |
|     56    | Customer#000000056 |              qh212iaGWtoVp               |      10     | 20-895-685-6920 |  6530.86  |  FURNITURE   |             quickly final dependencies. even dependencies are slyly regularly silent theodolites. slow a             |
|    104    | Customer#000000104 |     SEOogsfT y09vI2z PcSTnI18U6rNTf      |      10     | 20-966-284-8065 |  -588.38  |  FURNITURE   |                                           efully bold deposits. carefully                                            |
|    105    | Customer#000000105 |         XI8hMXfr8bIKTGhIRS2sYs,p         |      10     | 20-793-553-6417 |  9091.82  |  MACHINERY   |                                 solve pending, final requests. regular, bold platele                                 |
|    110    | Customer#000000110 |             FtEbLLBJQ24d2yG              |      10     | 20-893-536-2069 |  7462.99  |  AUTOMOBILE  |   unts. fluffily express dolphins cajole furiously. fluffily special pinto beans about the furiously express asym    |
|    158    | Customer#000000158 |       ZjQ2uJ,hZUZ,QCi,oNltA40 4dt        |      10     | 20-383-680-1329 |  6160.95  |  AUTOMOBILE  |                                        inal asymptotes. ironic accounts detec                                        |
|    159    | Customer#000000159 |      hCyzNocUmjEtz5xSYYzUCH1Bk1wF7       |      10     | 20-888-668-2668 |  2060.06  |  HOUSEHOLD   |      refully blithely final orbits. pending, final packages after the carefully ironic platelets nag carefully       |
|    166    | Customer#000000166 |   ,Vj4kzvCbTUmhLNSVjuyf8716NF5SddgDMY    |      10     | 20-320-530-5920 |  2042.21  |  FURNITURE   |      osits. regular, express deposits hang closely blithely regular platelets. unusual theodolites haggle fluf       |
|    175    | Customer#000000175 |   T2gZ1u7AC2Yve4DnuFWIzbeffURXkBn1PU7    |      10     | 20-427-617-9922 |  1975.35  |  FURNITURE   |                                         unts are slyly final foxes. always p                                         |
|    192    | Customer#000000192 |       AnEpYPRNZpiartUaRH7uTU2iTUn        |      10     | 20-750-712-2481 |  8239.96  |  MACHINERY   |                        

In [18]:
%%sql
-- all customer rows which have do not have c_nationkey as 10 or 20
SELECT
  *
FROM
  customer
WHERE
  c_nationkey NOT IN (10, 20);

Running query in 'duckdb:///tpch.db'

+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+----------------------------------------------------------------------------------------------------------------------+
| c_custkey |       c_name       |                c_address                 | c_nationkey |     c_phone     | c_acctbal | c_mktsegment |                                                      c_comment                                                       |
+-----------+--------------------+------------------------------------------+-------------+-----------------+-----------+--------------+----------------------------------------------------------------------------------------------------------------------+
|     1     | Customer#000000001 |            j5JsirBM9PsCy0O1m             |      15     | 25-989-741-2988 |   711.56  |   BUILDING   |                            y final requests wake slyly quickly special accounts. blithely                            |
|     2     | Customer#000000002 |      487LW1dovn6Q4dMVymKwwLE9OKf3QG      |      13     | 23-768-687-3665 |   121.65  |  AUTOMOBILE  |                           y carefully regular foxes. slyly regular requests about the bli                            |
|     3     | Customer#000000003 |               fkRGN8nY4pkE               |      1      | 11-719-748-3364 |  7498.12  |  AUTOMOBILE  |        fully. carefully silent instructions sleep alongside of the slyly regular asymptotes. quickly regular         |
|     4     | Customer#000000004 |               4u58h fqkyE                |      4      | 14-128-190-5944 |  2866.83  |  MACHINERY   |                                   sublate. fluffily even instructions are about th                                   |
|     5     | Customer#000000005 |       hwBtxkoBF qSW4KrIk5U 2B1AU7H       |      3      | 13-750-942-6364 |   794.47  |  HOUSEHOLD   |                                  equests haggle furiously against the pending packa                                  |
|     7     | Customer#000000007 |      8OkMVLQ1dK6Mbu6WG9 w4pLGQ n7MQ      |      18     | 28-190-982-9759 |  9561.95  |  AUTOMOBILE  |                    ounts. ironic, regular accounts sleep. final requests haggle quickly after the                    |
|     8     | Customer#000000008 |  j,pZ,Qp,qtFEo0r0c 92qobZtlhSuOqbE4JGV   |      17     | 27-147-574-9335 |  6819.74  |   BUILDING   |  riously final excuses sublate quickly among the fluffily even foxes. quickly final packages haggle furiously furi   |
|     9     | Customer#000000009 |        vgIql8H6zoyuLMFNdAMLyE7 H9        |      8      | 18-338-906-3675 |  8324.07  |  FURNITURE   |            ss pinto beans believe slyly quiet deposits-- doggedly bold packages boost. quickly ironic de             |
|     10    | Customer#000000010 |    Vf mQ6Ug9Ucf5OKGYq fsaX AtfsO7,rwY    |      5      | 15-741-346-9870 |  2753.54  |  HOUSEHOLD   |                                           g quickly after the evenly bold                                            |
|     11    | Customer#000000011 |         cG48rYjF3Aw7xs hKUXXqmI          |      23     | 33-464-151-3439 |  -272.60  |   BUILDING   |         ng to the regular foxes. furiously final deposits across the final platelets cajole quickly above th         |
|     12    | Customer#000000012 |              Sb4gxKs7W1AZa               |      13     | 23-791-276-1263 |  3396.49  |  HOUSEHOLD   |        ickly regular dependencies boost blithely around the slyly ironic theodolites. furiously special dolp         |
|     13    | Customer#000000013 |         Ez3ax0D5HnUbeUVSxoX8a8B          |      3      | 13-761-547-5974 |  3857.34  |   BUILDING   |          quickly brave foxes. blithely even packages against the pinto beans boost furiously against the re          |
|     14    | Customer#000000014 |             h3GFMzeFfYiamqr              |      1      | 11-845-129-3851 |  5266.30  |  FURNITURE   |                        

We can get the number of rows in a table using `count(*)` as shown below.


In [19]:
%%sql
SELECT
  COUNT(*)
FROM
  customer;

Running query in 'duckdb:///tpch.db'

+--------------+
| count_star() |
+--------------+
|    150000    |
+--------------+

In [20]:
%%sql
SELECT
  COUNT(*)
FROM
  lineitem;

Running query in 'duckdb:///tpch.db'

+--------------+
| count_star() |
+--------------+
|   6001215    |
+--------------+

If we want to get the rows sorted by values in a specific column, we use `ORDER BY`, for example.


In [21]:
%%sql
-- Will show the first ten customer records with the lowest custkey
-- rows are ordered in ASC order by default
SELECT
  *
FROM
  orders
ORDER BY
  o_custkey
LIMIT
  10;

Running query in 'duckdb:///tpch.db'

+------------+-----------+---------------+--------------+-------------+-----------------+-----------------+----------------+--------------------------------------------------------------------------------+
| o_orderkey | o_custkey | o_orderstatus | o_totalprice | o_orderdate | o_orderpriority |     o_clerk     | o_shippriority |                                   o_comment                                    |
+------------+-----------+---------------+--------------+-------------+-----------------+-----------------+----------------+--------------------------------------------------------------------------------+
|   454791   |     1     |       F       |   74602.81   |  1992-04-19 |     1-URGENT    | Clerk#000000815 |       0        |                   ccounts sleep carefully along the slyly ev                   |
|  4273923   |     1     |       O       |   95911.01   |  1997-03-23 |     3-MEDIUM    | Clerk#000000381 |       0        |                         he even deposits-- even, even                          |
|  5133509   |     1     |       O       |  174645.94   |  1996-07-01 |     1-URGENT    | Clerk#000000463 |       0        |                e across the furiously ironic deposits. furious                 |
|   579908   |     1     |       O       |   54048.26   |  1996-12-09 |      5-LOW      | Clerk#000000783 |       0        |  t packages hinder bold, even dolphins. slyly ironic packages wake fluffily a  |
|  3868359   |     1     |       F       |  123076.84   |  1992-08-22 |      5-LOW      | Clerk#000000536 |       0        |        ependencies boost blithely pending pinto beans. regular asymptot        |
|  4808192   |     1     |       O       |   65478.05   |  1996-06-29 |      2-HIGH     | Clerk#000000473 |       0        |  nag. final, special deposits believe carefully carefully express accounts. qu |
|   430243   |     2     |       F       |   33082.83   |  1994-12-24 | 4-NOT SPECIFIED | Clerk#000000121 |       0        |                e slyly along the furiously pending attainments                 |
|  1071617   |     2     |       P       |  221397.35   |  1995-03-10 |      2-HIGH     | Clerk#000000408 |       0        |         e carefully according to the furiously even packages. furious          |
|  1374019   |     2     |       F       |  167016.61   |  1992-04-05 |     1-URGENT    | Clerk#000000440 |       0        |                              uick, unusual courts                              |
|  2992930   |     2     |       F       |  103297.68   |  1994-05-21 | 4-NOT SPECIFIED | Clerk#000000913 |       0        |   even instructions affix slyly. furiously permanent decoys are furiously bli  |
+------------+-----------+---------------+--------------+-------------+-----------------+-----------------+----------------+--------------------------------------------------------------------------------+

In [22]:
%%sql
-- Will show the first ten customer's records with the highest custkey
SELECT
  *
FROM
  orders
ORDER BY
  o_custkey DESC
LIMIT
  10;

Running query in 'duckdb:///tpch.db'

+------------+-----------+---------------+--------------+-------------+-----------------+-----------------+----------------+-------------------------------------------------------------------+
| o_orderkey | o_custkey | o_orderstatus | o_totalprice | o_orderdate | o_orderpriority |     o_clerk     | o_shippriority |                             o_comment                             |
+------------+-----------+---------------+--------------+-------------+-----------------+-----------------+----------------+-------------------------------------------------------------------+
|  1070880   |   149999  |       O       |  126063.49   |  1995-12-02 |     3-MEDIUM    | Clerk#000000241 |       0        |   alongside of the slyly final accounts. ironic ideas are furious |
|  2488608   |   149999  |       F       |  139853.06   |  1994-10-07 |      2-HIGH     | Clerk#000000881 |       0        |                  even accounts may serve carefull                 |
|  2502080   |   149999  |       O       |  328596.74   |  1996-06-29 |     1-URGENT    | Clerk#000000207 |       0        |                       e. bold dolphins prin                       |
|  3775524   |   149999  |       O       |   11414.72   |  1995-06-09 |      2-HIGH     | Clerk#000000262 |       0        |     ly around the slyly final instructions. furiously special     |
|  4014148   |   149999  |       F       |  223396.15   |  1993-05-13 |      5-LOW      | Clerk#000000326 |       0        | forges sleep across the ironic requests. furiously express accoun |
|  4136421   |   149999  |       F       |  169716.79   |  1993-05-02 | 4-NOT SPECIFIED | Clerk#000000691 |       0        |      s. fluffily regular packages are carefully unusual theo      |
|  5200769   |   149999  |       O       |  163831.43   |  1995-05-09 |     1-URGENT    | Clerk#000000394 |       0        |     efully silent excuses. slyly express excuses after the pac    |
|  5323939   |   149999  |       F       |  212230.94   |  1994-03-05 |      2-HIGH     | Clerk#000000425 |       0        |           rint along the carefully bold dolphins. blith           |
|  5354048   |   149999  |       F       |   34669.55   |  1993-09-14 | 4-NOT SPECIFIED | Clerk#000000910 |       0        |         uriously above the careful deposits. theodolites d        |
|  5373249   |   149999  |       O       |   46217.79   |  1998-07-06 |      2-HIGH     | Clerk#000000744 |       0        |                      luffily carefully expre                      |
+------------+-----------+---------------+--------------+-------------+-----------------+-----------------+----------------+-------------------------------------------------------------------+


# [JOINS] Combine data from multiple tables using JOINs (there are different types of JOINs)

We can combine data from multiple tables using joins. When we write a join query, we have a format as shown below.

```sql
-- not based on real tables
SELECT
    a.*
FROM
    table_a a -- LEFT table a
    JOIN table_b b -- RIGHT table b
    ON a.id = b.id
```

The table specified first (table_a) is the left table, whereas the table established second is the right table. When we have multiple tables joined, we consider the joined dataset from the first two tables as the left table and the third table as the right table (The DB optimizes the joins for performance).

```sql
-- not based on real tables
SELECT
    a.*
FROM
    table_a a -- LEFT table a
    JOIN table_b b -- RIGHT table b
    ON a.id = b.id
    JOIN table_c c -- LEFT table is the joined data from 
-- table_a & table_b, right table is table_c
    ON a.c_id = c.id
```

There are five main types of joins, they are:

## 1. Inner join (default): Get only rows in both tables



In [23]:
%%sql
SELECT
  o.o_orderkey,
  l.l_orderkey
FROM
  orders o
  JOIN lineitem l ON o.o_orderkey = l.l_orderkey
  AND o.o_orderdate BETWEEN l.l_shipdate - INTERVAL '5' DAY AND l.l_shipdate  + INTERVAL '5' DAY
LIMIT
  100;

Running query in 'duckdb:///tpch.db'

+------------+------------+
| o_orderkey | l_orderkey |
+------------+------------+
|     7      |     7      |
|     32     |     32     |
|     33     |     33     |
|     69     |     69     |
|     71     |     71     |
|    133     |    133     |
|    198     |    198     |
|    259     |    259     |
|    260     |    260     |
|    288     |    288     |
|    353     |    353     |
|    354     |    354     |
|    420     |    420     |
|    482     |    482     |
|    485     |    485     |
|    486     |    486     |
|    512     |    512     |
|    517     |    517     |
|    548     |    548     |
|    576     |    576     |
|    576     |    576     |
|    581     |    581     |
|    614     |    614     |
|    678     |    678     |
|    679     |    679     |
|    737     |    737     |
|    771     |    771     |
|    771     |    771     |
|    772     |    772     |
|    774     |    774     |
|    899     |    899     |
|    962     |    962     |
|    962     |    962     |
|    992     |    992     |
|    992     |    992     |
|    1027    |    1027    |
|    1088    |    1088    |
|    1095    |    1095    |
|    1095    |    1095    |
|    1120    |    1120    |
|    1186    |    1186    |
|    1252    |    1252    |
|    1284    |    1284    |
|    1287    |    1287    |
|    1314    |    1314    |
|    1410    |    1410    |
|    1571    |    1571    |
|    1573    |    1573    |
|    1574    |    1574    |
|    1605    |    1605    |
|    1639    |    1639    |
|    1639    |    1639    |
|    1664    |    1664    |
|    1667    |    1667    |
|    1667    |    1667    |
|    1671    |    1671    |
|    1735    |    1735    |
|    1767    |    1767    |
|    1824    |    1824    |
|    1856    |    1856    |
|    1891    |    1891    |
|    1923    |    1923    |
|    1954    |    1954    |
|    1956    |    1956    |
|    1991    |    1991    |
|   122375   |   122375   |
|   122404   |   122404   |
|   122406   |   122406   |
|   122432   |   122432   |
|   122464   |   122464   |
|   122468   |   122468   |
|   122535   |   122535   |
|   122566   |   122566   |
|   122595   |   122595   |
|   122658   |   122658   |
|   122663   |   122663   |
|   122784   |   122784   |
|   122791   |   122791   |
|   122885   |   122885   |
|   122948   |   122948   |
|   122977   |   122977   |
|   122980   |   122980   |
|   123011   |   123011   |
|   123040   |   123040   |
|   123139   |   123139   |
|   123139   |   123139   |
|   123168   |   123168   |
|   123204   |   123204   |
|   123206   |   123206   |
|   123207   |   123207   |
|   123239   |   123239   |
|   123271   |   123271   |
|   123271   |   123271   |
|   123296   |   123296   |
|   123361   |   123361   |
|   123363   |   123363   |
|   123425   |   123425   |
|   123427   |   123427   |
|   123429   |   123429   |
|   123463   |   123463   |
+------------+------------+
Truncated to displaylimit of 100.

In [24]:
%%sql
SELECT
  COUNT(o.o_orderkey) AS order_rows_count,
  COUNT(l.l_orderkey) AS lineitem_rows_count
FROM
  orders o
  JOIN lineitem l ON o.o_orderkey = l.l_orderkey
  AND o.o_orderdate BETWEEN l.l_shipdate - INTERVAL '5' DAY AND l.l_shipdate  + INTERVAL '5' DAY;

Running query in 'duckdb:///tpch.db'

+------------------+---------------------+
| order_rows_count | lineitem_rows_count |
+------------------+---------------------+
|      247650      |        247650       |
+------------------+---------------------+

**Note:** JOIN defaults to INNER JOIN.

The output will have rows from orders and lineitem that found at least one matching row from the other table with the specified join condition (same o_orderkey and orderdate within ship date +/- 5 days). 

We can also see that 247650 rows from orders and lineitem tables matched.

## 2. Left outer join (aka left join): Get all rows from the left table and only matching rows from the right table.



In [25]:
%%sql
SELECT
  o.o_orderkey,
  l.l_orderkey
FROM
  orders o
  LEFT JOIN lineitem l ON o.o_orderkey = l.l_orderkey
  AND o.o_orderdate BETWEEN l.l_shipdate - INTERVAL '5' DAY AND l.l_shipdate  + INTERVAL '5' DAY
LIMIT
  100;

Running query in 'duckdb:///tpch.db'

+------------+------------+
| o_orderkey | l_orderkey |
+------------+------------+
|   122375   |   122375   |
|   122404   |   122404   |
|   122406   |   122406   |
|   122432   |   122432   |
|   122464   |   122464   |
|   122468   |   122468   |
|   122535   |   122535   |
|   122566   |   122566   |
|   122595   |   122595   |
|   122658   |   122658   |
|   122663   |   122663   |
|   122784   |   122784   |
|   122791   |   122791   |
|   122885   |   122885   |
|   122948   |   122948   |
|   122977   |   122977   |
|   123011   |   123011   |
|   123040   |   123040   |
|   123139   |   123139   |
|   123139   |   123139   |
|   123168   |   123168   |
|   123204   |   123204   |
|   123206   |   123206   |
|   123207   |   123207   |
|   123239   |   123239   |
|   123271   |   123271   |
|   123271   |   123271   |
|   123296   |   123296   |
|   123361   |   123361   |
|   123363   |   123363   |
|   123425   |   123425   |
|   123427   |   123427   |
|   123429   |   123429   |
|   123463   |   123463   |
|   123490   |   123490   |
|   123491   |   123491   |
|   123584   |   123584   |
|   123623   |   123623   |
|   123651   |   123651   |
|   123684   |   123684   |
|   123685   |   123685   |
|   123685   |   123685   |
|   123712   |   123712   |
|   123714   |   123714   |
|   123719   |   123719   |
|   123777   |   123777   |
|   123809   |   123809   |
|   123811   |   123811   |
|   123813   |   123813   |
|   123815   |   123815   |
|   123843   |   123843   |
|   123876   |   123876   |
|   123877   |   123877   |
|   123905   |   123905   |
|   123943   |   123943   |
|   123969   |   123969   |
|   124004   |   124004   |
|   124005   |   124005   |
|   124067   |   124067   |
|   124096   |   124096   |
|   124163   |   124163   |
|   124199   |   124199   |
|   124256   |   124256   |
|   124256   |   124256   |
|   124258   |   124258   |
|   124261   |   124261   |
|   124261   |   124261   |
|   124293   |   124293   |
|   124356   |   124356   |
|     7      |     7      |
|     32     |     32     |
|     33     |     33     |
|     69     |     69     |
|     71     |     71     |
|    133     |    133     |
|    198     |    198     |
|    259     |    259     |
|    260     |    260     |
|    288     |    288     |
|    353     |    353     |
|    354     |    354     |
|    420     |    420     |
|    482     |    482     |
|    485     |    485     |
|    486     |    486     |
|    512     |    512     |
|    517     |    517     |
|    548     |    548     |
|    576     |    576     |
|    576     |    576     |
|    581     |    581     |
|    614     |    614     |
|    678     |    678     |
|    679     |    679     |
|    737     |    737     |
|    771     |    771     |
|    771     |    771     |
|    772     |    772     |
|    774     |    774     |
|    899     |    899     |
+------------+------------+
Truncated to displaylimit of 100.

In [26]:
%%sql
SELECT
  COUNT(o.o_orderkey) AS order_rows_count,
  COUNT(l.l_orderkey) AS lineitem_rows_count
FROM
  orders o
  LEFT JOIN lineitem l ON o.o_orderkey = l.l_orderkey
  AND o.o_orderdate BETWEEN l.l_shipdate - INTERVAL '5' DAY AND l.l_shipdate  + INTERVAL '5' DAY;

Running query in 'duckdb:///tpch.db'

+------------------+---------------------+
| order_rows_count | lineitem_rows_count |
+------------------+---------------------+
|     1519332      |        247650       |
+------------------+---------------------+

The output will have all the rows from orders and the rows from lineitem that were able to find at least one matching row from the orders table with the specified join condition (same o_orderkey and orderdate within ship date +/- 5 days). 

We can also see that the number of rows from the orders table is 1,519,332 & from the lineitem table is 247,650. The number of rows in orders is 1,500,000, but the join condition produces 1,519,332 since some orders match with multiple lineitems.

## 3. Right outer join (aka right join): Get matching rows from the left and all rows from the right table.



In [27]:
%%sql
SELECT
  o.o_orderkey,
  l.l_orderkey
FROM
  orders o
  RIGHT JOIN lineitem l ON o.o_orderkey = l.l_orderkey
  AND o.o_orderdate BETWEEN l.l_shipdate - INTERVAL '5' DAY AND l.l_shipdate  + INTERVAL '5' DAY
LIMIT
  100;

Running query in 'duckdb:///tpch.db'

+------------+------------+
| o_orderkey | l_orderkey |
+------------+------------+
|     7      |     7      |
|     32     |     32     |
|     33     |     33     |
|     69     |     69     |
|     71     |     71     |
|    133     |    133     |
|    198     |    198     |
|    259     |    259     |
|    260     |    260     |
|    288     |    288     |
|    353     |    353     |
|    354     |    354     |
|    420     |    420     |
|    482     |    482     |
|    485     |    485     |
|    486     |    486     |
|    512     |    512     |
|    517     |    517     |
|    548     |    548     |
|    576     |    576     |
|    576     |    576     |
|    581     |    581     |
|    614     |    614     |
|    678     |    678     |
|    679     |    679     |
|    737     |    737     |
|    771     |    771     |
|    771     |    771     |
|    772     |    772     |
|    774     |    774     |
|    899     |    899     |
|    962     |    962     |
|    962     |    962     |
|    992     |    992     |
|    992     |    992     |
|    1027    |    1027    |
|    1088    |    1088    |
|    1095    |    1095    |
|    1095    |    1095    |
|    1120    |    1120    |
|    1186    |    1186    |
|    1252    |    1252    |
|    1284    |    1284    |
|    1287    |    1287    |
|    1314    |    1314    |
|    1410    |    1410    |
|    1506    |    1506    |
|    1506    |    1506    |
|    1571    |    1571    |
|    1573    |    1573    |
|    1574    |    1574    |
|    1605    |    1605    |
|    1639    |    1639    |
|    1639    |    1639    |
|    1664    |    1664    |
|    1667    |    1667    |
|    1667    |    1667    |
|    1671    |    1671    |
|    1735    |    1735    |
|    1767    |    1767    |
|    1794    |    1794    |
|    1824    |    1824    |
|    1825    |    1825    |
|    1856    |    1856    |
|    1891    |    1891    |
|    1923    |    1923    |
|    1954    |    1954    |
|    1956    |    1956    |
|    1991    |    1991    |
|   122375   |   122375   |
|   122404   |   122404   |
|   122406   |   122406   |
|   122432   |   122432   |
|   122464   |   122464   |
|   122468   |   122468   |
|   122535   |   122535   |
|   122566   |   122566   |
|   122595   |   122595   |
|   122658   |   122658   |
|   122663   |   122663   |
|   122784   |   122784   |
|   122791   |   122791   |
|   122885   |   122885   |
|   122948   |   122948   |
|   122977   |   122977   |
|   122980   |   122980   |
|   123011   |   123011   |
|   123040   |   123040   |
|   123139   |   123139   |
|   123139   |   123139   |
|   123168   |   123168   |
|   123204   |   123204   |
|   123206   |   123206   |
|   123207   |   123207   |
|   123239   |   123239   |
|   123271   |   123271   |
|   123271   |   123271   |
|   123296   |   123296   |
|   123361   |   123361   |
|   123363   |   123363   |
+------------+------------+
Truncated to displaylimit of 100.

In [28]:
%%sql
SELECT
  COUNT(o.o_orderkey) AS order_rows_count,
  COUNT(l.l_orderkey) AS lineitem_rows_count
FROM
  orders o
  JOIN lineitem l ON o.o_orderkey = l.l_orderkey
  AND o.o_orderdate BETWEEN l.l_shipdate - INTERVAL '5' DAY AND l.l_shipdate  + INTERVAL '5' DAY;

Running query in 'duckdb:///tpch.db'

+------------------+---------------------+
| order_rows_count | lineitem_rows_count |
+------------------+---------------------+
|      247650      |        247650       |
+------------------+---------------------+

The output will have the rows from orders that found at least one matching row from the lineitem table with the specified join condition (same o_orderkey and orderdate within ship date +/- 5 days) and all the rows from the lineitem table.

We can also see that the number of rows from the orders table is 247,650 & from the lineitem table is 6,001,215 .

## 4. Full outer join: Get all rows from both the left and right tables.



In [29]:
%%sql
SELECT
  o.o_orderkey,
  l.l_orderkey
FROM
  orders o
  FULL OUTER JOIN lineitem l ON o.o_orderkey = l.l_orderkey
  AND o.o_orderdate BETWEEN l.l_shipdate - INTERVAL '5' DAY AND l.l_shipdate  + INTERVAL '5' DAY
LIMIT
  100;

Running query in 'duckdb:///tpch.db'

+------------+------------+
| o_orderkey | l_orderkey |
+------------+------------+
|   122375   |   122375   |
|   122404   |   122404   |
|   122406   |   122406   |
|   122432   |   122432   |
|   122464   |   122464   |
|   122468   |   122468   |
|   122535   |   122535   |
|   122566   |   122566   |
|   122595   |   122595   |
|   122658   |   122658   |
|   122663   |   122663   |
|   122691   |   122691   |
|   122784   |   122784   |
|   122791   |   122791   |
|   122885   |   122885   |
|   122948   |   122948   |
|   122977   |   122977   |
|   122980   |   122980   |
|   123011   |   123011   |
|   123040   |   123040   |
|   123139   |   123139   |
|   123139   |   123139   |
|   123168   |   123168   |
|   123204   |   123204   |
|   123206   |   123206   |
|   123207   |   123207   |
|   123239   |   123239   |
|   123271   |   123271   |
|   123271   |   123271   |
|   123296   |   123296   |
|   123361   |   123361   |
|   123363   |   123363   |
|   123425   |   123425   |
|   123427   |   123427   |
|   123429   |   123429   |
|   123463   |   123463   |
|   123490   |   123490   |
|   123491   |   123491   |
|   123584   |   123584   |
|   123623   |   123623   |
|   123651   |   123651   |
|   123684   |   123684   |
|   123685   |   123685   |
|   123685   |   123685   |
|   123712   |   123712   |
|   123714   |   123714   |
|   123719   |   123719   |
|   123777   |   123777   |
|   123809   |   123809   |
|   123811   |   123811   |
|   123813   |   123813   |
|   123815   |   123815   |
|   123843   |   123843   |
|   123876   |   123876   |
|   123877   |   123877   |
|   123905   |   123905   |
|   123943   |   123943   |
|   123969   |   123969   |
|   124004   |   124004   |
|   124005   |   124005   |
|   124067   |   124067   |
|   124096   |   124096   |
|   124163   |   124163   |
|   124199   |   124199   |
|   124256   |   124256   |
|   124256   |   124256   |
|   124258   |   124258   |
|   124261   |   124261   |
|   124261   |   124261   |
|   124293   |   124293   |
|   124356   |   124356   |
|     7      |     7      |
|     32     |     32     |
|     33     |     33     |
|     69     |     69     |
|     71     |     71     |
|    133     |    133     |
|    198     |    198     |
|    259     |    259     |
|    260     |    260     |
|    288     |    288     |
|    353     |    353     |
|    354     |    354     |
|    420     |    420     |
|    482     |    482     |
|    485     |    485     |
|    486     |    486     |
|    512     |    512     |
|    517     |    517     |
|    548     |    548     |
|    576     |    576     |
|    576     |    576     |
|    581     |    581     |
|    614     |    614     |
|    678     |    678     |
|    679     |    679     |
|    737     |    737     |
|    771     |    771     |
|    771     |    771     |
|    772     |    772     |
+------------+------------+
Truncated to displaylimit of 100.

In [30]:
%%sql
SELECT
  COUNT(o.o_orderkey) AS order_rows_count,
  COUNT(l.l_orderkey) AS lineitem_rows_count
FROM
  orders o
  FULL OUTER JOIN lineitem l ON o.o_orderkey = l.l_orderkey
  AND o.o_orderdate BETWEEN l.l_shipdate - INTERVAL '5' DAY AND l.l_shipdate  + INTERVAL '5' DAY;

Running query in 'duckdb:///tpch.db'

+------------------+---------------------+
| order_rows_count | lineitem_rows_count |
+------------------+---------------------+
|     1519332      |       6001215       |
+------------------+---------------------+

The output will have all the rows from orders that found at least one matching row from the lineitem table with the specified join condition (same o_orderkey and orderdate within ship date +/- 5 days) and all the rows from the lineitem table.

We can also see that the number of rows from the orders table is 1,519,332 & from the lineitem table is 6,001,215.

## 5. Cross join: Get the cartesian product of all rows



In [31]:
%%sql
SELECT
  n.n_name AS nation_c_name,
  r.r_name AS region_c_name
FROM
  nation n
  CROSS JOIN region r;

Running query in 'duckdb:///tpch.db'

+----------------+---------------+
| nation_c_name  | region_c_name |
+----------------+---------------+
|    ALGERIA     |     AFRICA    |
|   ARGENTINA    |     AFRICA    |
|     BRAZIL     |     AFRICA    |
|     CANADA     |     AFRICA    |
|     EGYPT      |     AFRICA    |
|    ETHIOPIA    |     AFRICA    |
|     FRANCE     |     AFRICA    |
|    GERMANY     |     AFRICA    |
|     INDIA      |     AFRICA    |
|   INDONESIA    |     AFRICA    |
|      IRAN      |     AFRICA    |
|      IRAQ      |     AFRICA    |
|     JAPAN      |     AFRICA    |
|     JORDAN     |     AFRICA    |
|     KENYA      |     AFRICA    |
|    MOROCCO     |     AFRICA    |
|   MOZAMBIQUE   |     AFRICA    |
|      PERU      |     AFRICA    |
|     CHINA      |     AFRICA    |
|    ROMANIA     |     AFRICA    |
|  SAUDI ARABIA  |     AFRICA    |
|    VIETNAM     |     AFRICA    |
|     RUSSIA     |     AFRICA    |
| UNITED KINGDOM |     AFRICA    |
| UNITED STATES  |     AFRICA    |
|    ALGERIA     |    AMERICA    |
|   ARGENTINA    |    AMERICA    |
|     BRAZIL     |    AMERICA    |
|     CANADA     |    AMERICA    |
|     EGYPT      |    AMERICA    |
|    ETHIOPIA    |    AMERICA    |
|     FRANCE     |    AMERICA    |
|    GERMANY     |    AMERICA    |
|     INDIA      |    AMERICA    |
|   INDONESIA    |    AMERICA    |
|      IRAN      |    AMERICA    |
|      IRAQ      |    AMERICA    |
|     JAPAN      |    AMERICA    |
|     JORDAN     |    AMERICA    |
|     KENYA      |    AMERICA    |
|    MOROCCO     |    AMERICA    |
|   MOZAMBIQUE   |    AMERICA    |
|      PERU      |    AMERICA    |
|     CHINA      |    AMERICA    |
|    ROMANIA     |    AMERICA    |
|  SAUDI ARABIA  |    AMERICA    |
|    VIETNAM     |    AMERICA    |
|     RUSSIA     |    AMERICA    |
| UNITED KINGDOM |    AMERICA    |
| UNITED STATES  |    AMERICA    |
|    ALGERIA     |      ASIA     |
|   ARGENTINA    |      ASIA     |
|     BRAZIL     |      ASIA     |
|     CANADA     |      ASIA     |
|     EGYPT      |      ASIA     |
|    ETHIOPIA    |      ASIA     |
|     FRANCE     |      ASIA     |
|    GERMANY     |      ASIA     |
|     INDIA      |      ASIA     |
|   INDONESIA    |      ASIA     |
|      IRAN      |      ASIA     |
|      IRAQ      |      ASIA     |
|     JAPAN      |      ASIA     |
|     JORDAN     |      ASIA     |
|     KENYA      |      ASIA     |
|    MOROCCO     |      ASIA     |
|   MOZAMBIQUE   |      ASIA     |
|      PERU      |      ASIA     |
|     CHINA      |      ASIA     |
|    ROMANIA     |      ASIA     |
|  SAUDI ARABIA  |      ASIA     |
|    VIETNAM     |      ASIA     |
|     RUSSIA     |      ASIA     |
| UNITED KINGDOM |      ASIA     |
| UNITED STATES  |      ASIA     |
|    ALGERIA     |     EUROPE    |
|   ARGENTINA    |     EUROPE    |
|     BRAZIL     |     EUROPE    |
|     CANADA     |     EUROPE    |
|     EGYPT      |     EUROPE    |
|    ETHIOPIA    |     EUROPE    |
|     FRANCE     |     EUROPE    |
|    GERMANY     |     EUROPE    |
|     INDIA      |     EUROPE    |
|   INDONESIA    |     EUROPE    |
|      IRAN      |     EUROPE    |
|      IRAQ      |     EUROPE    |
|     JAPAN      |     EUROPE    |
|     JORDAN     |     EUROPE    |
|     KENYA      |     EUROPE    |
|    MOROCCO     |     EUROPE    |
|   MOZAMBIQUE   |     EUROPE    |
|      PERU      |     EUROPE    |
|     CHINA      |     EUROPE    |
|    ROMANIA     |     EUROPE    |
|  SAUDI ARABIA  |     EUROPE    |
|    VIETNAM     |     EUROPE    |
|     RUSSIA     |     EUROPE    |
| UNITED KINGDOM |     EUROPE    |
| UNITED STATES  |     EUROPE    |
+----------------+---------------+
Truncated to displaylimit of 100.


The output will have every row of the nation joined with every row of the region. There are 25 nations and five regions, leading to 125 rows in our result from the cross-join.

![Joins](./images/im_27_1.png)
![Cross join](./images/im_27_2.png)

There are cases where we will need to join a table with itself, called a SELF-join.

**Example**:

1. For every customer order, get the order placed earlier in the same week (Sunday - Saturday, not the previous seven days). Only show customer orders that have at least one such order.



In [32]:
%%sql
SELECT
  o1.o_custkey
FROM
  orders o1
  JOIN orders o2 ON o1.o_custkey = o2.o_custkey
  AND YEAR (o1.o_orderdate) = YEAR (o2.o_orderdate)
  AND week (o1.o_orderdate) = week (o2.o_orderdate)
WHERE
  o1.o_orderkey != o2.o_orderkey;

Running query in 'duckdb:///tpch.db'

+-----------+
| o_custkey |
+-----------+
|   130057  |
|    9220   |
|   93385   |
|   123823  |
|   55126   |
|   39907   |
|   50417   |
|   11704   |
|   64754   |
|   38299   |
|   134617  |
|   31696   |
|    3478   |
|   116326  |
|   108556  |
|   105941  |
|   28924   |
|   133400  |
|    4306   |
|   65312   |
|   29749   |
|   38071   |
|   44923   |
|   99505   |
|   36427   |
|   71026   |
|   18637   |
|   54376   |
|   61123   |
|   73696   |
|   77266   |
|    5663   |
|   48157   |
|    4243   |
|   92359   |
|   111901  |
|   33227   |
|   88393   |
|   101821  |
|   77359   |
|   87323   |
|   20941   |
|   31168   |
|   104617  |
|   103651  |
|   24016   |
|   121615  |
|   66976   |
|   60367   |
|   13177   |
|   98308   |
|   38041   |
|   27061   |
|   80107   |
|   63397   |
|   121156  |
|   68575   |
|   17677   |
|   138236  |
|   44716   |
|   100228  |
|   22951   |
|   46567   |
|   89615   |
|    7828   |
|   147937  |
|   98675   |
|   148414  |
|   93235   |
|   53362   |
|   15377   |
|   56425   |
|   110447  |
|   83902   |
|   26557   |
|   33688   |
|   130417  |
|   123316  |
|   108506  |
|   101696  |
|   23443   |
|   22462   |
|   13658   |
|   106831  |
|   122150  |
|   135751  |
|   49093   |
|   23536   |
|   16915   |
|   110323  |
|   77701   |
|   144865  |
|    997    |
|   83569   |
|   79765   |
|   38032   |
|   84449   |
|   22543   |
|   60352   |
|   35525   |
+-----------+
Truncated to displaylimit of 100.


Most analytical queries require calculating metrics that involve combining data from multiple rows. `GROUP BY` allows us to perform aggregate calculations on data from a set of rows grouped together based on values of specified column(s).

**Example**:

1. Create a report that shows the number of orders per orderpriority segment.



In [33]:
%%sql
SELECT
  o_orderpriority,
  COUNT(*) AS num_orders
FROM
  orders
GROUP BY
  o_orderpriority;

Running query in 'duckdb:///tpch.db'

+-----------------+------------+
| o_orderpriority | num_orders |
+-----------------+------------+
|     1-URGENT    |   300343   |
|      2-HIGH     |   300091   |
|     3-MEDIUM    |   298723   |
|      5-LOW      |   300589   |
| 4-NOT SPECIFIED |   300254   |
+-----------------+------------+


In the above query, we group the data by `orderpriority`, and the calculation `count(*)` will be applied to the rows having a specific `orderpriority` value. The output will consist of one row per unique value of `orderpriority` and the `count(*)` calculation.

![Group by](./images/im_28.png)

The calculations allowed are typically SUM/MIN/MAX/AVG/COUNT. However, some databases have more complex aggregate functions; check your DB documentation.

# [SUB QUERY] Use the result of a query within a query using sub-queries

When we want to use the result of a query as a table in another query, we use subqueries. **Example**:

1. Create a report that shows the nation, how many items it supplied (by suppliers in that nation), and how many items it purchased (by customers in that nation). 



In [34]:
%%sql
SELECT
  n.n_name AS nation_c_name,
  s.quantity AS supplied_items_quantity,
  c.quantity AS purchased_items_quantity
FROM
  nation n
  LEFT JOIN (
    SELECT
      n.n_nationkey,
      SUM(l.l_quantity) AS quantity
    FROM
      lineitem l
      JOIN supplier s ON l.l_suppkey = s.s_suppkey
      JOIN nation n ON s.s_nationkey = n.n_nationkey
    GROUP BY
      n.n_nationkey
  ) s ON n.n_nationkey = s.n_nationkey
  LEFT JOIN (
    SELECT
      n.n_nationkey,
      SUM(l.l_quantity) AS quantity
    FROM
      lineitem l
      JOIN orders o ON l.l_orderkey = o.o_orderkey
      JOIN customer c ON o.o_custkey = c.c_custkey
      JOIN nation n ON c.c_nationkey = n.n_nationkey
    GROUP BY
      n.n_nationkey
  ) c ON n.n_nationkey = c.n_nationkey;

Running query in 'duckdb:///tpch.db'

+----------------+-------------------------+--------------------------+
| nation_c_name  | supplied_items_quantity | purchased_items_quantity |
+----------------+-------------------------+--------------------------+
|    ALGERIA     |        6454691.00       |        6117618.00        |
|   ARGENTINA    |        6339724.00       |        6087566.00        |
|     BRAZIL     |        6085551.00       |        6149174.00        |
|     CANADA     |        6296547.00       |        6168913.00        |
|     EGYPT      |        6385468.00       |        6024134.00        |
|    ETHIOPIA    |        5817697.00       |        6095241.00        |
|     FRANCE     |        6141618.00       |        6289987.00        |
|    GERMANY     |        6076474.00       |        6098776.00        |
|     INDIA      |        6347392.00       |        6102406.00        |
|   INDONESIA    |        6204759.00       |        6276420.00        |
|      IRAN      |        5992033.00       |        6091271.00        |
|      IRAQ      |        6678487.00       |        6003025.00        |
|     JAPAN      |        5775308.00       |        6067924.00        |
|     JORDAN     |        5539782.00       |        6230581.00        |
|     KENYA      |        5753471.00       |        6002935.00        |
|    MOROCCO     |        5691519.00       |        6060121.00        |
|   MOZAMBIQUE   |        6253676.00       |        6243695.00        |
|      PERU      |        6412494.00       |        6035782.00        |
|     CHINA      |        6205793.00       |        6179322.00        |
|    ROMANIA     |        6079467.00       |        6212841.00        |
|  SAUDI ARABIA  |        6299506.00       |        5948921.00        |
|    VIETNAM     |        6111369.00       |        6149667.00        |
|     RUSSIA     |        6152008.00       |        6265778.00        |
| UNITED KINGDOM |        5946859.00       |        6046145.00        |
| UNITED STATES  |        6037102.00       |        6130552.00        |
+----------------+-------------------------+--------------------------+


In the above query, we can see that there are two sub-queries, one to calculate the quantity supplied by a nation and the other to calculate the quantity purchased by the customers of a nation.

# [CASE WHEN] Replicate IF.ELSE logic with CASE statements

We can do conditional logic in the `SELECT ... FROM` part of our query, as shown below.



In [35]:
%%sql
SELECT
    o_orderkey,
    o_totalprice,
    CASE
        WHEN o_totalprice > 100000 THEN 'high'
        WHEN o_totalprice BETWEEN 25000
        AND 100000 THEN 'medium'
        ELSE 'low'
    END AS order_price_bucket
FROM
    orders;

Running query in 'duckdb:///tpch.db'

+------------+--------------+--------------------+
| o_orderkey | o_totalprice | order_price_bucket |
+------------+--------------+--------------------+
|     1      |  173665.47   |        high        |
|     2      |   46929.18   |       medium       |
|     3      |  193846.25   |        high        |
|     4      |   32151.78   |       medium       |
|     5      |  144659.20   |        high        |
|     6      |   58749.59   |       medium       |
|     7      |  252004.18   |        high        |
|     32     |  208660.75   |        high        |
|     33     |  163243.98   |        high        |
|     34     |   58949.67   |       medium       |
|     35     |  253724.56   |        high        |
|     36     |   68289.96   |       medium       |
|     37     |  206680.66   |        high        |
|     38     |   82500.05   |       medium       |
|     39     |  341734.47   |        high        |
|     64     |   39414.99   |       medium       |
|     65     |  110643.60   |        high        |
|     66     |  103740.67   |        high        |
|     67     |  169405.01   |        high        |
|     68     |  330793.52   |        high        |
|     69     |  197689.49   |        high        |
|     70     |  113534.42   |        high        |
|     71     |  276992.74   |        high        |
|     96     |   68989.90   |       medium       |
|     97     |  110512.84   |        high        |
|     98     |   69168.33   |       medium       |
|     99     |  112126.95   |        high        |
|    100     |  187782.63   |        high        |
|    101     |  124906.11   |        high        |
|    102     |  164529.10   |        high        |
|    103     |  126990.79   |        high        |
|    128     |   66195.16   |       medium       |
|    129     |  261013.14   |        high        |
|    130     |  189484.12   |        high        |
|    131     |  130464.09   |        high        |
|    132     |  144947.21   |        high        |
|    133     |  114663.57   |        high        |
|    134     |  200354.30   |        high        |
|    135     |  213713.99   |        high        |
|    160     |  124132.06   |        high        |
|    161     |   34439.93   |       medium       |
|    162     |   2726.45    |        low         |
|    163     |  202379.28   |        high        |
|    164     |  301925.76   |        high        |
|    165     |  188023.67   |        high        |
|    166     |  154451.92   |        high        |
|    167     |   71124.35   |       medium       |
|    192     |  167668.42   |        high        |
|    193     |   88699.17   |       medium       |
|    194     |  176707.84   |        high        |
|    195     |  191542.31   |        high        |
|    196     |   47095.31   |       medium       |
|    197     |  163700.57   |        high        |
|    198     |  176189.31   |        high        |
|    199     |  112986.49   |        high        |
|    224     |  259960.36   |        high        |
|    225     |  220441.09   |        high        |
|    226     |  313134.62   |        high        |
|    227     |   54880.58   |       medium       |
|    228     |   2749.87    |        low         |
|    229     |  213076.69   |        high        |
|    230     |  171318.37   |        high        |
|    231     |  191808.17   |        high        |
|    256     |  130775.45   |        high        |
|    257     |   8649.81    |        low         |
|    258     |  294182.42   |        high        |
|    259     |  123298.34   |        high        |
|    260     |  272353.52   |        high        |
|    261     |  319306.86   |        high        |
|    262     |  143146.21   |        high        |
|    263     |  149369.68   |        high        |
|    288     |  266047.60   |        high        |
|    289     |  211162.52   |        high        |
|    290     |   99019.42   |       medium       |
|    291     |  108550.11   |        high        |
|    292     |   40637

We can see how we display different values depending on the `totalprice` column. We can also use multiple criteria as our conditional criteria (e.g., totalprice > 100000 AND orderpriority = '2-HIGH').

# [STANDARD FUNCTION] Use standard inbuilt DB functions for common String, Time, and Numeric data manipulation

When processing data, more often than not, we will need to change values in columns; shown below are a few standard functions to be aware of:

1. **String functions**
   1. **LENGTH** is used to calculate the length of a string. E.g., `SELECT LENGTH('hi');` will output 2.
   2. **CONCAT** combines multiple string columns into one. E.g., `SELECT CONCAT(o_orderstatus, '-', o_orderpriority) FROM ORDERS LIMIT 5;` will concatenate the o_orderstatus and o_orderpriority columns with a dash in between them.
   3. **SPLIT** is used to split a value into an array based on a given delimiter. E.g., `SELECT STRING_SPLIT(o_orderpriority, '-') FROM ORDERS LIMIT 5;` will output a column with arrays formed by splitting o_orderpriority values on `-`.
   4. **SUBSTRING** is used to get a sub-string from a value, given the start and end character indices. E.g., `SELECT o_orderpriority, SUBSTRING(o_orderpriority, 1, 5) FROM orders LIMIT 5;` will get the first five (1 - 5) characters of the o_orderpriority column.
   5. **TRIM** is used to remove empty spaces to the left and right of the value. E.g., `SELECT TRIM(' hi ');` will output `hi` without any spaces around it. LTRIM and RTRIM are similar but only remove spaces before and after the string, respectively.
2. **Date and Time functions**
   1. **Adding and subtracting dates**: Is used to add and subtract periods; the format heavily depends on the DB. In DuckDB, `date_diff` accepts 3 parameters, the outputs unit (day, month, year), the datetime/date values a and b such that the output will be a - b. The `+ INTERVAL n UNIT(DAY/MONTH/YEAR)` adds the value (in specified units) to the timestamp value.
      ```sql
        -- Date and Time Functions
        SELECT
            date_diff('day', DATE '2022-10-01', DATE '2023-11-05') AS diff_in_days,
            date_diff('month', DATE '2022-10-01', DATE '2023-11-05') AS diff_in_months,
            date_diff('year', DATE '2022-10-01', DATE '2023-11-05') AS diff_in_years,
            DATE '2022-10-01' + INTERVAL 400 DAY AS new_date;
      ```
    It will show the difference between the two dates in the specified period. We can also add/subtract an arbitrary period from a date/time column. E.g., `SELECT DATE '2022-11-05' + INTERVAL '10' DAY;` will show the output `2022-11-15` (try subtraction of dates).
   2. **String <=> date/time conversions**: When we want to change the data type of a string to date/time, we can use the `DATE 'YYYY-MM-DD'` or `TIMESTAMP 'YYYY-MM-DD HH:mm:SS'` functions. But when the data is in a non-standard date/time format such as `MM/DD/YYYY`, we will need to specify the input structure; we do this using `strptime`, E.g., `SELECT strptime('11-05-2023', '%m-%d-%Y');`.  
   
   We can convert a timestamp/date into a string with the required format using `strftime`. E.g., `SELECT strftime(o_orderdate, '%Y-%m-01') AS first_month_date FROM orders LIMIT 5;` will map every o_orderdate to the first of their month.

   See [this page](https://duckdb.org/docs/sql/functions/date) on how to set the proper date time format.
   3. **Time frame functions (YEAR/MONTH/DAY)**:  When we want to extract specific periods from a date/time column, we can use these functions. E.g., `SELECT year(DATE '2023-11-05');` will return 2023. Similarly, we have month, day, hour, min, etc.
3. **Numeric**
   1. **ROUND** is used to specify the number of digits allowed after the decimal point. E.g. `SELECT ROUND(100.102345, 2);`
   2. **ABS** is used to get the absolute value of a given number. E.g. `SELECT ABS(-100), ABS(100);` 
   3. **Mathematical operations** these are +,-,*,/.
   4. **Ceil/Floor** is used to get the next higher and most recent lower integers, given a decimal digit. E.g. `SELECT CEIL(100.1), FLOOR(100.1);`